### ISSUE TITLE
file system linking errors on Debian 10 with packaged DUNE 2.6 version and CUDA

### ISSUE THREAD

**Issue by blattms** [OPM/opm-simulators#2363]
I might be doing something terribly wrong when compiling OPM with dune 2.6 on Debian 10, but I am getting [linker errors about the filesystem stuff](https://github.com/OPM/opm-simulators/files/4245337/compile.log).

`-lstdc++fs` is provided in the linker line, but not picked up. If I add the full path of the static library it
works.

BTW there is a typo at [OpmCompilerChecks.cmake#L96](https://github.com/OPM/opm-common/blob/fb75bcd4e25d64caa3b940c7f4c717e23bdc0ac8/cmake/Modules/OpmCompilerChecks.cmake#L96). But that check seems to be bogus anyway. I am using a g++ version where according to the fixed test the lib is not needed but it actually is:
```
mblatt@smaug:~/src/dune/opm-2.6$ c++ --version
c++ (Debian 8.3.0-6) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
mblatt@smaug:~/src/dune/opm-2.6$ cmake --version
cmake version 3.13.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

~~BTW I am requesting shared libs explicitly with `-DBUILD_SHARED_LIBS=ON`. Still this is somehow ignored by our buildsystem as it seems to have built libopmcommon.a. I will test whether that is the issue.~~

**Comment by blattms** [comment]
CMake options are:
```
set(USE_MPI OFF         CACHE BOOL "Use mpi")
set(CMAKE_DISABLE_FIND_PACKAGE_MPI TRUE CACHE BOOL "disable mpi")
set(BUILD_TESTING ON CACHE BOOL "Whether to build tests")
set(CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY 1 CACHE BOOL "" FORCE)
set(CMAKE_DISABLE_FIND_PACKAGE_OpenMP ON CACHE BOOL "dispable OpenMP")
```

**Comment by blattms** [comment]
Hence I do not request shared libs and building static ones seems ok.

**Comment by blattms** [comment]
Could be something special on my system. CMake is issuing some suspicuous warnings:
```
CMake Warning at /home/mblatt/src/dune/opm-2.6/opm-common/cmake/Modules/OpmSatellites.cmake:60 (add_executable):
  Cannot generate a safe linker search path for target test_timer because
  files in some directories may conflict with libraries in implicit
  directories:

    link library [libcublas.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs
    link library [libcusparse.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs

  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  /home/mblatt/src/dune/opm-2.6/opm-common/cmake/Modules/OpmLibMain.cmake:299 (opm_compile_satellites)
  CMakeLists.txt:121 (include)
```

**Comment by blattms** [comment]
That looks a bit messed up:
```
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ dpkg --listfiles nvidia-cuda-dev| grep libcublas
/usr/lib/x86_64-linux-gnu/libcublas_device.a
/usr/lib/x86_64-linux-gnu/libcublas_static.a
/usr/lib/x86_64-linux-gnu/stubs/libcublas.so
/usr/lib/x86_64-linux-gnu/libcublas.so
/usr/share/man/man7/libcublas.7.gz
/usr/share/man/man7/libcublas.so.7.gz
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ ls -l /usr/lib/x86_64-linux-gnu/stubs/libcublas.so
-rw-r--r-- 1 root root 53472 Jun 13  2018 /usr/lib/x86_64-linux-gnu/stubs/libcublas.so
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ ls -l /usr/lib/x86_64-linux-gnu/libcublas.so
lrwxrwxrwx 1 root root 16 Mär 22  2019 /usr/lib/x86_64-linux-gnu/libcublas.so -> libcublas.so.9.2
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ ls -l /usr/lib/x86_64-linux-gnu/libcublas.so.9.2
lrwxrwxrwx 1 root root 20 Mär 22  2019 /usr/lib/x86_64-linux-gnu/libcublas.so.9.2 -> libcublas.so.9.2.174
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ ls -l /usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174 
-rw-r--r-- 1 root root 62641232 Jul 14  2018 /usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ 
```

**Comment by blattms** [comment]
Same picture with cuda deactivated.

**Comment by blattms** [comment]
Sigh, If I use `-DBUILD_SHARED_LIBS=ON` everything works fine. Of course the installed DUNE only provides shared libs. Hence my guess is that the linker somehow gets confused.

I also blame my special aura for me being the only one having this problem.

Like I said, if I explcitly add the static library to link command line then it works. The problem is that it is locate at `/usr/lib/gcc/x86_64-linux-gnu/8/libstdc++fs.a` and hence I would need to do some ugly hacks to find it. The location probably differs form distribution to distribution.

@akva2, do you have an idea how to resolve this cleanly?

**Comment by akva2** [comment]
your setup is exactly the same as jenkins, and there it works fine, so i am really confused why you are seeing issues :/ 

have you tried -Wl,-Bstatic -lstdc++fs -Wl,-Bdynamic ?

**Comment by blattms** [comment]
And why is the problem only apparent in opm-simulators?

**Comment by blattms** [comment]
Nope that does not help.

**Comment by akva2** [comment]
show us the linker line? (make VERBOSE=1)

**Comment by blattms** [comment]
```
/usr/bin/cmake -E cmake_link_script CMakeFiles/test_timer.dir/link.txt --verbose=1
/usr/bin/c++  -pipe -Wall -Wextra -Wshadow -pthread -O3 -DNDEBUG -mtune=native  -rdynamic -Wl,--enable-new-dtags CMakeFiles/test_timer.dir/tests/test_timer.cpp.o CMakeFiles/test_timer.dir/cmake_device_link.o  -o bin/test_timer  -L/home/mblatt/src/dune/opm-2.6/opm-common/opm-seq/lib  -L/home/mblatt/src/dune/opm-2.6/opm-grid/opm-seq/lib  -L/usr/lib/x86_64-linux-gnu/stubs  -L/usr/lib/gcc/x86_64-linux-gnu/7 -Wl,-rpath,/home/mblatt/src/dune/opm-2.6/opm-common/opm-seq/lib:/home/mblatt/src/dune/opm-2.6/opm-grid/opm-seq/lib:/usr/lib/x86_64-linux-gnu/openmpi/lib lib/libopmsimulators.a /home/mblatt/src/dune/opm-2.6/opm-grid/opm-seq/lib/libopmgrid.a -lboost_date_time /usr/lib/x86_64-linux-gnu/libdunealbertagrid_1d.so.2.6-git /usr/lib/x86_64-linux-gnu/libdunealbertagrid_2d.so.2.6-git /usr/lib/x86_64-linux-gnu/libdunealbertagrid_3d.so.2.6-git -lalberta_1d -lalberta_2d -lalberta_3d -lalberta_utilities -lltdl -lumfpack -lamd -llapack -lblas -lm -lsuitesparseconfig -lrt /home/mblatt/src/dune/opm-2.6/opm-common/opm-seq/lib/libopmcommon.a -lstdc++fs -lpthread -lboost_system -ltrilinos_zoltan -lscotch -lptscotch -lptscotcherr -lboost_unit_test_framework /usr/lib/x86_64-linux-gnu/libdunegrid.so.2.6-git -lm -lltdl /usr/lib/x86_64-linux-gnu/libdunegeometry.so.2.6-git /usr/lib/x86_64-linux-gnu/libugS2.so.2.6.0 /usr/lib/x86_64-linux-gnu/libugS3.so.2.6.0 /usr/lib/x86_64-linux-gnu/libugL.so.2.6.0 -pthread /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so /usr/lib/x86_64-linux-gnu/libdunecommon.so.2.6-git -lcublas -lcusparse -lcudadevrt -lcudart_static -lrt -lpthread -ldl 
```

**Comment by blattms** [comment]
Wondering where `-L/usr/lib/gcc/x86_64-linux-gnu/7` comes from. That is the wrong compiler version. It should be 8 must be dragged in by some dependency.

**Comment by akva2** [comment]
yes. it looks to be dragged in by opm-grid.

**Comment by blattms** [comment]
Yes, this is a clean build.
I tend to concur with Linus when it come to NVIDIA. Dragged in  by nvidia-cuda-toolkit which does not use the default system compiler, but gcc-7.
On the other hand, I did try a build with cuda deactivated and that did not change the picture.

**Comment by blattms** [comment]
I will try to deactivate dendencies and see what happens.

**Comment by blattms** [comment]
Indeed, it is cuda that breaks compilation. I had a typo in the CMake options when I first tried.

**Comment by blattms** [comment]
With cuda CMake also prints a ton of warnings like:
```

CMake Warning at /home/mblatt/src/dune/opm-2.6/opm-common/cmake/Modules/OpmSatellites.cmake:300 (add_executable):
  Cannot generate a safe linker search path for target ebos because files in
  some directories may conflict with libraries in implicit directories:

    link library [libcublas.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs
    link library [libcusparse.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs

  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  CMakeLists.txt:252 (opm_add_test)


CMake Warning at /home/mblatt/src/dune/opm-2.6/opm-common/cmake/Modules/OpmSatellites.cmake:60 (add_executable):
  Cannot generate a safe linker search path for target
  test_multmatrixtransposed because files in some directories may conflict
  with libraries in implicit directories:

    link library [libcublas.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs
    link library [libcusparse.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs

  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  /home/mblatt/src/dune/opm-2.6/opm-common/cmake/Modules/OpmLibMain.cmake:299 (opm_compile_satellites)
  CMakeLists.txt:121 (include)


CMake Warning at /home/mblatt/src/dune/opm-2.6/opm-common/cmake/Modules/OpmSatellites.cmake:300 (add_executable):
  Cannot generate a safe linker search path for target ebos_polymer because
  files in some directories may conflict with libraries in implicit
  directories:

    link library [libcublas.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs
    link library [libcusparse.so] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /usr/lib/x86_64-linux-gnu/stubs

```

Somehow there are multiple cuda versions of the cuda libraries on the system:
```
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ locate libcublas.so
/usr/lib/x86_64-linux-gnu/libcublas.so
/usr/lib/x86_64-linux-gnu/libcublas.so.9.2
/usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
/usr/lib/x86_64-linux-gnu/stubs/libcublas.so
/usr/share/man/man7/libcublas.so.7.gz
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ readlink -f /usr/lib/x86_64-linux-gnu/libcublas.so
/usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ ls -l /usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
-rw-r--r-- 1 root root 62641232 Jul 14  2018 /usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ ls -l /usr/lib/x86_64-linux-gnu/stubs/libcublas.so
-rw-r--r-- 1 root root 53472 Jun 13  2018 /usr/lib/x86_64-linux-gnu/stubs/libcublas.so
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ dpkg -S usr/lib/x86_64-linux-gnu/stubs/libcublas.so
nvidia-cuda-dev: /usr/lib/x86_64-linux-gnu/stubs/libcublas.so
mblatt@smaug:~/src/dune/opm-2.6/opm-simulators/opm-seq$ dpkg -S  /usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
libcublas9.2:amd64: /usr/lib/x86_64-linux-gnu/libcublas.so.9.2.174
```
nvidia-cuda-dev depends on libcublas9.2. Interestingly, CMakeCache.txt uses the libs in /usr/lib/x86_64-linux-gnu/ with absolute path and we pass them to `target_link_libraries`. Yet the linker command uses -lcublas and the likes and neglects the path.

**Comment by blattms** [comment]
Might be resolvable by using NEW of [CMP0060](https://cmake.org/cmake/help/v3.3/policy/CMP0060.html#policy:CMP0060).

**Comment by blattms** [comment]
No that had nothing to do with it.

If found a fix for CMake >=3.8. We did not support lower versions after the CUDA stuff was added anyway. We do with the fix again but then we deactivate CUDA:
